File: algoritmos_de_ordenação/merge.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void mergesort(long int vet[],long int inicio,long int fim);
void ordena(long int vet[],long int inicio,long int meio, long int fim);
long int * aloca_vetor_randomico(long int qtd);

int main(){
    long int * vet;
    long int tam;

    
    printf("\n Digite o tamanho do vetor: ");
    scanf("%ld",&tam);


    vet = aloca_vetor_randomico(tam);
    
    mergesort(vet,0,tam-1);
    
    
    return 0;
}


void mergesort(long int vet[],long int inicio,long int fim){
    long int meio;
    

    if(inicio<fim){
        meio = (inicio + fim) /2;
        mergesort(vet,inicio,meio);
        mergesort(vet,meio+1,fim);
        ordena(vet,inicio,meio,fim);
    }
    
    return ;
}

void ordena(long int vet[],long int inicio,long int meio, long int fim){
    long int i,j,k,*aux;

    aux=(long int*)calloc(sizeof(long int), fim - inicio + 1);
    i=inicio;
    j=meio+1;
    k=0;

    
    while(i <= meio && j <= fim){
        if(vet[i]<vet[j]){
            aux[k] = vet[i];
            i++;
        }else
        {
            aux[k] = vet[j];
            j++;
        }
        k++;
    }
    
    while(i<=meio){
        aux[k] = vet[i];
        i++;
        k++;
    }
    while(j<=meio){
        aux[k] = vet[j];
        j++;
        k++;
    }

    for(i=0;i<(fim-inicio)+1;i++){
        vet[inicio+1] = aux[i];
    }
    
    free(aux);
    return;
}
long int * aloca_vetor_randomico(long int qtd)
{
    long int * vetor;
    long int i;
    long int i_aux, aux;
    vetor = (long int*)malloc(sizeof(long int)*qtd);

    for(i=0;i<qtd;i++)
    {
        vetor[i] = i+1;
    }
    srand(time(NULL));

    for(i=0;i<qtd;i++)
    {
        i_aux = rand() % qtd;
        aux = vetor[i];
        vetor[i] = vetor[i_aux];
        vetor[i_aux] = aux;

    }

    return vetor;
}

File: algoritmos_de_ordenação/merge.py
list = [54,26,93,17,77,31,44,55,20]

def Mergesort(vetor):
    aux = vetor[:]
    mergesort2(vetor, aux, 0, len(vetor) - 1)


def mergesort2(vetor, aux, start, end):
    if start >= end:
        return

    meio = (start + end) // 2

    mergesort2(vetor, aux, start, meio)
    mergesort2(vetor, aux, meio + 1, end)

    ordena(vetor, aux, start, end)


def ordena(vetor, aux, start, end):
    i = start
    ifim = (start + end) // 2

    j = ifim + 1
    jfim = end

    for p in range(start, end + 1):
        #  se não há elementos a esquerda
        if i > ifim:
            aux[p] = vetor[j]
            j += 1

        #  se não há elementos a direita
        elif j > jfim:
            aux[p] = vetor[i]
            i += 1

        #  compara esquerda com direita
        elif vetor[i] < vetor[j]:
            aux[p] = vetor[i]
            i += 1

        else:
            aux[p] = vetor[j]
            j += 1

    for k in range(start, end + 1):
        vetor[k] = aux[k]
